lib/plugin/instance.rb

```diff
@@ -615,18 +622,30 @@ def register_locales!
         Rails.configuration.assets.precompile << "locales/#{locale}.js"
       else
         msg = "Invalid locale! #{opts.inspect}"
-        # The logger isn't always present during boot / parsing locales from plugins
-        if Rails.logger.present?
-          Rails.logger.error(msg)
-        else
-          puts msg
-        end
+        log_error(msg)
       end
     end
   end
 
   private
 
+  # When booting/parsing locales from plugins, some actions are invoked before Rails logger.
```

Comment:
Feels to me like it makes more sense to sort out boot order here so we have access to a logger. 

If we have no chance here we should log to STDERR vs STDOUT for warnings and errors

Reply:
That make sense. This is actually a peak of iceberg. I think we should make this module much cleaner since it's plugin API. Some ideas:

1. We probably need a way to document where a plugin method is in Rails booting sequence.
2. And we put all our logic to somewhere else so API is clear for reading.
3. It might be worthy to to versionize API too.

I think the problem is that the module is really hard to test due to weird into booting sequence.

Reply:
> Feels to me like it makes more sense to sort out boot order here so we have access to a logger.

This is quite tricky... all the auth providers need to be registered in time for the omniauth initializer. Therefore auth providers are registered super early on, at the `activate!` stage:
 https://github.com/discourse/discourse/blob/master/config/application.rb#L245

The plugin `after_initialize` hook is too late.

We can add a specific hook for 'pre-omniauth', or we could do something like this so that `Discourse.deprecate` saves its warning for when the application has initialized:

```
@@ -613,7 +616,11 @@ module Discourse
     redis_key = "deprecate-notice-#{digest}"
 
     if !$redis.without_namespace.get(redis_key)
-      Rails.logger.warn(warning)
+      if Rails.initialized?
+        Rails.logger.warn(warning)
+      else
+        Rails.application.config.after_initialize { Rails.logger.warn(warning) }
+      end
       $redis.without_namespace.setex(redis_key, 3600, "x")
     end
```
^^ This option would allow us to use Discourse.deprecate in places where the app hasn't yet loaded, which could be useful. At the same time, it feels kinda hacky...

What do you think @SamSaffron?

Reply:
> This is quite tricky... all the auth providers need to be registered in time for the omniauth initializer. Therefore auth providers are registered super early on, at the activate! stage:
Can we move activation to a initializer? We have logger there.

1. Hopefully it doesn't mess up seed/assets managing
2. Putting it before `009-omniauth.rb` should fix the problem

Reply:
Yep, we can definitely do that. It could even be done inside the 009-omniauth initializer. I was trying to decide whether to 'fix' `Discourse.deprecate`, or work around it by adding this extra hook to the initializer 🤔 

Now that a few hours have passed I think the initializer is the better option, as you said @fantasticfears. I'm sure I'll change my mind again by tomorrow morning 😆

Reply:
Well, this approach comes a lot of work because plugin hooks into every aspects of Discourse. When I started to look at auth module, I didn't know the existence of `Discourse.deprecate`. The good thing is that it's only used in `active_record_base.rb` and `sql_builder.rb`.

Happy hacking :P You have the power to design.